### PR TITLE
Separate unshard stream for each process group with FSDP

### PR DESCRIPTION
**Pull request by snarayan21**
This PR gives each `_FSDPState` with a separate process group its own unshard stream. This addresses an issue we have been seeing with overlapping communication and computation when doing 2D parallelism with FSDP.

Essentially, the whole model is sharded across all ranks, and communication is done over the default process group which contains all ranks. We 2D parallelize the model by sharding a subset of the parameters over distinct process groups. For example, with 16 GPUs, most of the model is sharded across all 16 GPUs, but some parameters are sharded across groups of 4 GPUs, and replicated 4 times. So GPU 0 will be part of the default process group, but also another process group that includes ranks 0, 1, 2, 3. For these parameters, all-gathers and reduce-scatters are done with this new process group. We FSDP wrap the parent class with the default process group, but also wrap these parameters with the new process group.

The issue arises because FSDP normally sets the same unshard stream for all `_FSDPState` objects, meaning that the default computation stream will wait on any communication kernels that are currently in the unshard stream. As far as I understand, with just one process group, this is necessary because we need to wait for parameters to be all-gathered before proceeding with computation. However, with >1 process group, we find that computation will wait for parameters to be unsharded even when those parameters are not needed for that computation. In both the forwards and backwards passes, this results in the computation stream excessively waiting for all-gathers that do not need to be waited on.

This PR has a solution to the above issue. For each process group attached to any `_FSDPState`, we assign a new unshard stream. With this, the computation stream only waits on all-gather ops that are actually necessary for that computation to proceed. With 2D parallelism, computation that requires unsharding parameters from all ranks (using the default process group) will only wait on all-gathers that use the default process group, while computation that requires unsharding parameters from ranks in the custom process group will only wait on all-gathers from the custom process group. This eliminates bubbles where computation and communication overlap, and according to our tests, improves training efficiency by up to 25% at scale.

There are definitely many ways to address this so we wanted to get input on how to best solve this, and how this can be addressed in the upcoming FSDP rewrite and in support for 2D parallelism. Given the discussions [here](https://github.com/NVIDIA/nccl/issues/195) and [here](https://github.com/NVIDIA/nccl/issues/195) there may be some risks involved in the current approach with using multiple NCCL communicators concurrently, since each process group may have its own NCCL communicator (unsure about this though). FWIW in our testing we have not seen any hangs and training has been identical before and after the change. It also may be possible to FSDP wrap our model differently to make sure that communications are queued at the right time.

Looking forward to discussing this further and helping with a fix!

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -53,6 +53,36 @@ class _PrefetchMode(Enum):
...
+    """
+    from torch.distributed.distributed_c10d import get_process_group_ranks
+    if state.process_group is None:
+        # If no process group is attached to the _FSDPState, assume it uses default process group.
+        return True
+    return len(get_process_group_ranks(
+        state.process_group)) == dist.get_world_size()
+
+
+def fsdp_state_pg_ranks(state: _FSDPState) -> Tuple[int]:
```
Comment (Skylion007) [inline torch/distributed/fsdp/_runtime_utils.py:74]:
```suggestion
def fsdp_state_pg_ranks(state: _FSDPState) -> Tuple[int, ...]:
```
is probably what you actually mean.

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Please fix your submodules an make sure to run lintrunner on this PR.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -213,23 +250,25 @@ def _share_state_and_init_handle_attrs(
...
+        if hasattr(flat_param, '_in_backward_optimizers'):
             raise RuntimeError(
-                "FSDP optimizer in backward only supported with use_orig_params=True!"
+                'FSDP optimizer in backward only supported with use_orig_params=True!'
             )
         handle._has_optim_in_backward = flat_param._params is not None and any(
-            hasattr(param, "_in_backward_optimizers") for param in flat_param._params
-        )
-        if handle._has_optim_in_backward:
-            torch._C._log_api_usage_once("fsdp.optimizer_in_backward")
```
Comment (Skylion007) [inline torch/distributed/fsdp/_runtime_utils.py:227]: Did you mean to delete this log_api_usage call?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -200,11 +230,14 @@ def _share_state_and_init_handle_attrs(
...
-    ``root_module`` 's module tree, and initializes handle attributes. These
-    are done together to require a single loop over the states.
-    """
+    """Shares data structure state from the ``root_state`` to all FSDP states in
+    ``root_module`` 's module tree, and initializes handle attributes. These are
+    done together to require a single loop over the states."""
+    from torch.distributed.fsdp._runtime_utils import (
+        HOMOGENEOUS_ATTR_NAMES, _init_device_mesh,
+        _validate_and_get_hybrid_shard_state)
+    from torch.distributed.utils import _p_assert
```
Comment (Skylion007) [inline torch/distributed/fsdp/_runtime_utils.py:239]: This is already in the global namescope of the python file? Any reason to reimport it?

Comment (snarayan21) [reply torch/distributed/fsdp/_runtime_utils.py:239]: oop just copy-pasted from my patch. Will remove this

File: torch/distributed/fsdp/_runtime_utils.py [hunk 4]
```diff
@@ -53,6 +54,37 @@ class _PrefetchMode(Enum):
     FORWARD = auto()
 
 
+def fsdp_state_has_default_pg(state: _FSDPState) -> bool:
+    """Indicates whether FlatParamHandle has the default process group.
+
+    Args:
+        handle (_FSDPState): FSDP State object
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:61]: nit: This looks like it has a typo. I would be good with not including this comment at all since the code seems self-explanatory.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 5]
```diff
@@ -53,6 +54,37 @@ class _PrefetchMode(Enum):
...
+            otherwise.
+    """
+    if state.process_group is None:
+        # If no process group is attached to the _FSDPState, assume it uses default process group.
+        return True
+    return len(get_process_group_ranks(state.process_group)) == dist.get_world_size()
+
+
+def fsdp_state_pg_ranks(state: _FSDPState) -> Tuple[int, ...]:
+    """Gets the ranks included in the ProcessGroup of an _FSDPState.
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:74]: nit: Same with this comment -- I am good with not including it.

**Review by awgu — DISMISSED** [review]
This sounds good to me!

I recall that the concurrent NCCL communicator issues was resolved in a recent NCCL update? cc: @kwen2501

**Comment by awgu** [comment]
Upon a closer look, I think the unit test failures are real.

I wonder if they can be fixed by changing
https://github.com/pytorch/pytorch/blob/f543093e065f4b248dc8c96936ae4062fc1382b0/torch/distributed/fsdp/_runtime_utils.py#L604-L608
such that every unshard stream waits on the default stream (i.e. current stream).

**Comment by snarayan21** [comment]
Looks like there was one issue with not being able to access the `priority` of a `Stream`, so addressed that. I also added in a `_wait_for_computation_stream` call for each of the unshard streams. Hopefully this addresses the FSDP-DDP discrepancy

File: torch/distributed/fsdp/_runtime_utils.py (hunk 4 above)
Comment (snarayan21) [reply torch/distributed/fsdp/_runtime_utils.py]: done!

File: torch/distributed/fsdp/_runtime_utils.py (hunk 5 above)
Comment (snarayan21) [reply torch/distributed/fsdp/_runtime_utils.py]: done!

**Comment by snarayan21** [comment]
Looks like one test for a model is not passing but I'm not sure if that's deterministic or not?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 6]
```diff
@@ -225,6 +240,13 @@ def _share_state_and_init_handle_attrs(
         )
         if handle._has_optim_in_backward:
             torch._C._log_api_usage_once("fsdp.optimizer_in_backward")
+    # Keep track of any new unshard streams we may have to add for specific process groups.
+    fsdp_pg_unshard_streams = {}
+    try:
+        unshard_priority = root_state._unshard_stream.priority
```
Comment (Skylion007) [inline torch/distributed/fsdp/_runtime_utils.py:246]: Can we just add a priority attribute to the Stream class and initialize it to 0 or something??

File: torch/distributed/fsdp/_runtime_utils.py [hunk 7]
```diff
@@ -253,6 +293,13 @@ def _share_state_and_init_handle_attrs(
         handle = fsdp_state._handle
         if handle:
             handle.init_flat_param_attributes()
+    # Ensure that all unshard streams wait on the default computation stream
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:296]: Wait sorry, I think this needs to be done in the root pre-forward so that we call this every iteration. I think `_share_state_and_init_handle_attrs` should only run once.

Here is where the wait stream happens in `_root_pre_forward`:
https://github.com/pytorch/pytorch/blob/2b2b3c453660bf57543dc3420007c5ccf1317809/torch/distributed/fsdp/_runtime_utils.py#L646-L650

I think ideally, we should refactor `_wait_for_computation_stream()` to take in the root state and inside the function, we get all unshard streams in all FSDP states, the reason being that there is still only one pre-unshard stream shared across all FSDP states.

Comment (snarayan21) [reply torch/distributed/fsdp/_runtime_utils.py:296]: Ah I see, yes this makes sense to do. I thought `_lazy_init`, and by extension, `_share_state_and_init_handle_attrs`, would run in every call to `_root_pre_forward`. But didn't realize it's a no-op after the first call to forward. I'll get this change in now.

**Comment by snarayan21** [comment]
@awgu The test linked here was failing because a `Stream` object didn't have the `priority` attribute: https://github.com/pytorch/pytorch/actions/runs/7389134737/job/20102917057?pr=116611

**Comment by awgu** [comment]
@snarayan21 Thanks for the pointer!

Can we change:
https://github.com/pytorch/pytorch/blob/7073dc604e220e1690dc21c77b52da1e63a534e1/torch/cpu/__init__.py#L55-L61

**Comment by snarayan21** [comment]
Fixed the dummy `Stream` class `priority` attribute issue in this commit.

**Comment by snarayan21** [comment]
> I recall that the concurrent NCCL communicator issues was resolved in a recent NCCL update?
@awgu We are currently seeing some issues with deadlocks when using multiple process groups. Do you have more information on what change to NCCL resolved this, and maybe a NCCL version? That would be super helpful, thanks!

File: torch/distributed/fsdp/_runtime_utils.py (hunk 6 above)
Comment (snarayan21) [reply torch/distributed/fsdp/_runtime_utils.py:246]: resolved by change below

**Review by awgu — DISMISSED** [review]
LGTM!

Regarding the concurrent NCCL PGs, I am not too sure. @kwen2501 do you have any insights?

**Comment by awgu** [comment]
@snarayan21 There is still a unit test failing 🤔 
```
python test/distributed/fsdp/test_fsdp_core.py -k test_mixture_of_experts_offload_false_none
```

I might need to find some time to patch your PR and look into this one. I do not have a sense of why it is failing off the top of my head.

**Comment by snarayan21** [comment]
@awgu one datapoint for debugging this unit test failure: seems to work for the first iteration but not the second.

**Comment by wconstab** [comment]
> Given the discussions https://github.com/NVIDIA/nccl/issues/195 and https://github.com/NVIDIA/nccl/issues/195 there may be some risks involved in the current approach with using multiple NCCL communicators concurrently, since each process group may have its own NCCL communicator (unsure about this though)

It should be OK to use multiple NCCL communicators for unshard operations (allgathers), provided you ensure that the allgathers go in the same order on every rank.  If you don't ensure this, you could observe an allgather timeout.

**Comment by snarayan21** [comment]
Running into an issues with non-deterministic loss curves with this patch. Am working on a fix and will update the PR when that happens.

**Comment by snarayan21** [comment]
@awgu I'm still looking into how to address this problem of computation stream waiting on all-gather operations that the computation actually doesn't require. Is there a way to address this without using multiple unshard streams? From what I can tell, there's no way to make `wait_stream` conditional, as in, only have computation of type A wait on all-gathers of type A, and only have computation of type B wait on all-gathers of type B, without using separate streams. I thought about using `events` but it seems that the `wait_stream` call uses `record_event` under the hood as well. My understanding of this is incomplete, so is there possibly a way to use waiting on particular events to solve this problem?

**Comment by awgu** [comment]
@snarayan21 Using CUDA events with a single unshard stream may not address this issue because waiting on an event from `stream` will wait for any pending work in `stream` from before the event was recorded, which means that the CPU thread's issue order matters.

Were you able to understand better why the fix in the PR is insufficient?

**Comment by awgu** [comment]
For per-parameter sharding, we are using CUDA events whenever possible to avoid the heavier-weight CUDA stream synchronization. However, as mentioned in the previous comment, waiting on an event will still wait for all pending work issued before that event was recorded.

One question I have is, for your use case, does the order in which all-gathers are issued by the CPU match the order in which the all-gathers are expected to be waited on by the computation stream?

**Comment by snarayan21** [comment]
@awgu Yeah the more I think about it and investigate, seems like the issue order is partially incorrect. But even after correcting the all-gather issue order in the backwards pass, this issue persists. I'm going to look into redoing our FSDP wrapping and see if that addresses the issue we are seeing, hopefully improving efficiency and giving deterministic loss curves.

**Comment by wconstab** [comment]
> looks like the all-gather issue order is partially incorrect (in the backwards pass due to the _post_forwards calls order being wrong). 

we dont have much control over the order that autograd's engine traverses the dag of read backward ops.  if fsdp encounters a case like 

b = foo(a); c = bar(a)

where foo and bar are both FSDP-wrapped modules that depend on a common input, I wonder how we deal with ensuring the backwards ordering is consistent between ranks?  It's valid for autograd to pick either ordering of dFoo, dBar or dBar, dFoo since they are peers in the graph.  This leads to 2 subproblems potentially-
1) you actually care that dBar goes first.  (bc thats more optimal from a comms overlap perspective)
2) we need to have the same order on every rank, even if we don't care about (1) - we can't let rank0 and rank1 pick different orders

Do we have a solution for (2) today in FSDP or are we just not hitting this case in real models?  @awgu

**Comment by snarayan21** [comment]
Seems that we were able to resolve the core issue from our side by modifying how we do FSDP wrapping for particular modules. In light of that, making multiple unshard streams might be overkill. I can close out this issue unless people want to continue the discussion. Thanks yall! :)

**Comment by awgu** [comment]
@snarayan21 Sounds good. Let us close this for now then.
